coderd/workspaceapps.go

```diff
@@ -119,3 +226,113 @@ func (api *API) workspaceAppsProxyPath(rw http.ResponseWriter, r *http.Request)
+func (a ApplicationURL) String() string {
+	appNameOrPort := a.AppName
+	if a.Port != 0 {
+		appNameOrPort = strconv.Itoa(int(a.Port))
+	}
+
+	return fmt.Sprintf("%s--%s--%s--%s.%s", appNameOrPort, a.AgentName, a.WorkspaceName, a.Username, a.BaseHostname)
+}
+
+// ParseSubdomainAppURL parses an application from the subdomain of r's Host
+// header. If the subdomain is not a valid application URL hostname, returns a
+// non-nil error.
+//
+// Subdomains should be in the form:
+//
+//	{PORT/APP_NAME}--{AGENT_NAME}--{WORKSPACE_NAME}--{USERNAME}
+//	(eg. http://8080--main--dev--dean.hi.c8s.io)
+func ParseSubdomainAppURL(hostname string) (ApplicationURL, error) {
+	subdomain, rest, err := SplitSubdomain(hostname)
+	if err != nil {
+		return ApplicationURL{}, xerrors.Errorf("split host domain %q: %w", hostname, err)
+	}
+
+	matches := appURL.FindAllStringSubmatch(subdomain, -1)
+	if len(matches) == 0 {
+		return ApplicationURL{}, xerrors.Errorf("invalid application url format: %q", subdomain)
+	}
+	matchGroup := matches[0]
+
+	appName, port := AppNameOrPort(matchGroup[appURL.SubexpIndex("AppName")])
+	return ApplicationURL{
+		AppName:       appName,
+		Port:          port,
+		AgentName:     matchGroup[appURL.SubexpIndex("AgentName")],
+		WorkspaceName: matchGroup[appURL.SubexpIndex("WorkspaceName")],
+		Username:      matchGroup[appURL.SubexpIndex("Username")],
+		BaseHostname:  rest,
+	}, nil
+}
+
+// SplitSubdomain splits a subdomain from the rest of the hostname. E.g.:
+//   - "foo.bar.com" becomes "foo", "bar.com"
+//   - "foo.bar.baz.com" becomes "foo", "bar.baz.com"
+//
+// An error is returned if the string doesn't contain a period.
+func SplitSubdomain(hostname string) (subdomain string, rest string, err error) {
+	toks := strings.SplitN(hostname, ".", 2)
+	if len(toks) < 2 {
+		return "", "", xerrors.New("no subdomain")
+	}
+
+	return toks[0], toks[1], nil
+}
+
+// AppNameOrPort takes a string and returns either the input string or a port
+// number.
+func AppNameOrPort(val string) (string, uint16) {
+	port, err := strconv.ParseUint(val, 10, 16)
+	if err != nil || port == 0 {
+		port = 0
+	} else {
+		val = ""
+	}
+
+	return val, uint16(port)
+}
+
+// applicationCookie is a helper function to copy the auth cookie to also
+// support subdomains. Until we support creating authentication cookies that can
+// only do application authentication, we will just reuse the original token.
+// This code should be temporary and be replaced with something that creates
+// a unique session_token.
+//
+// Returns nil if the access URL isn't a hostname.
+func (api *API) applicationCookie(authCookie *http.Cookie) *http.Cookie {
+	if net.ParseIP(api.AccessURL.Hostname()) != nil {
+		return nil
+	}
+
+	appCookie := *authCookie
+	// We only support setting this cookie on the access URL subdomains. This is
+	// to ensure we don't accidentally leak the auth cookie to subdomains on
+	// another hostname.
+	appCookie.Domain = "." + api.AccessURL.Hostname()
```

Comment:
This is interesting... I didn't know we could apply to subdomains without requiring an authentication redirect, this is awesome.

Reply:
We probably won't be doing this soon as we'll be implementing proper devurl auth similarly to v1

Reply:
@kylecarbs this is a stop gap. It will not work on apps hosted on another domain. 

Also currently using the same full auth token. We should used scoped auth tokens.